add post (#194)
rstudio/ai-blog: _posts/2022-10-18-torch-linalg/torch-linalg.Rmd

---
title: "Five ways to do least squares (with torch)"
description: >

>   Get to know torch's linalg module, all while learning about different ways to do least-squares regression from scratch. This post, again, is a condensed version of the corresponding chapter in the forthcoming book, Deep Learning and Scientific Computing with R torch, to be published by CRC Press. 

Reviewer: ```suggestion
  Get to know torch's linalg module, all while learning about different ways to do least-squares regression from scratch. This post is a condensed version of the corresponding chapter in the forthcoming book, Deep Learning and Scientific Computing with R torch, to be published by CRC Press. 
```

author:
  - name: Sigrid Keydana
    affiliation: RStudio
    affiliation_url: https://www.rstudio.com/
slug: keydanatorchlinalg
date: 2022-10-18
categories:
  - Torch
  - R
  - Concepts
  - Tabular data
output:
  distill::distill_article:
    self_contained: false
    toc: true
preview: images/squirrel.jpg
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE, fig.width = 6, fig.height = 6)
```

> *Note: This post is a condensed version of a chapter from part three of the forthcoming book, Deep Learning and Scientific Computing with R torch. Part three is dedicated to scientific computation beyond deep learning. As throughout the book, but -- if possible -- to an even higher degree, there I focus on the underlying concepts, striving to explain them in as "verbal" a way as I can. This does not mean skipping the equations; it means taking care to explain why they are the way they are.*

Reviewer: > As throughout the book, but -- if possible -- to an even higher degree, there I focus on the underlying concepts

You've lost me a little bit here with the caveats. To a higher degree than what? I think you might just simplify this to "Throughout the book I focus on underlying concepts…” Are you trying to say that in the book, as opposed to in this post, you focus on underlying concepts? If that's the case, I'd state that in as straightforward a manner as possible. 🙂

Author: no, I meant in part 3 even more than parts 1 & 2 ... you're right, I'll just throw it out completely!

How do you compute linear least-squares regression? In R, using `lm()`; in `torch`, there is `linalg_lstsq()`.

> Where R, sometimes, hides complexity from the user, high-performance computation frameworks like `torch` tend to ask a bit more up-front effort, be it careful reading of documentation, or playing around some, or both. For example, here is the central piece of documentation for `linalg_lstsq()`, elaborating on the `driver` parameter to the function:

Reviewer: ```suggestion
Where R, sometimes, hides complexity from the user, high-performance computation frameworks like `torch` tend to ask for a bit more effort up front, be it careful reading of documentation, or playing around some, or both. For example, here is the central piece of documentation for `linalg_lstsq()`, elaborating on the `driver` parameter to the function:
```

```
`driver` chooses the LAPACK/MAGMA function that will be used.
For CPU inputs the valid values are 'gels', 'gelsy', 'gelsd, 'gelss'.
For CUDA input, the only valid driver is 'gels', which assumes that A is full-rank.
To choose the best driver on CPU consider:
  -   If A is well-conditioned (its condition number is not too large), or you do not mind some precision loss:
     -   For a general matrix: 'gelsy' (QR with pivoting) (default)
     -   If A is full-rank: 'gels' (QR)
  -   If A is not well-conditioned:
     -   'gelsd' (tridiagonal reduction and SVD)
     -   But if you run into memory issues: 'gelss' (full SVD).
```

> Whether you'll need to know this will depend on the problem you're solving. But if you do, it certainly will help to have an idea what is being talked about there, if only in a high-level way.

Reviewer: ```suggestion
Whether you'll need to know this will depend on the problem you're solving. But if you do, it certainly will help to have an idea what is being described there, if only in a high-level way.
```

Reviewer: Described, or discussed, or something of the like. For some reason “talked about” always sounds odd to me when referencing something non-verbal in writing.

Author: yeah, right! I think "alluded to" will do fine ...

## The plan

The way we'll organize this exploration is by solving a least-squares problem from scratch, making use of various matrix factorizations. Concretely, we'll approach the task:

1.  By means of the so-called *normal equations*, the most direct way, in the sense that it immediately results from a mathematical statement of the problem.

2.  Again, starting from the normal equations, but making use of *Cholesky factorization* in solving them.

3.  Yet again, taking the normal equations for a point of departure, but proceeding by means of *LU* decomposition.

4.  Fourth, employing another type of factorization -- *QR* -- that, together with the final one, accounts for the vast majority of decompositions applied "in the real world". With QR decomposition, the solution algorithm does not start from the normal equations.

> 5.  And fifth and finally, making use of *Singular Value Decomposition* (SVD). Here, too, the normal equations are not needed.

Reviewer: I don't think you need to say "fourth" and "fifth" given that it's a numbered list.

Author: Hm, not quite sure what exactly you mean, but looking at it, I see that only (4) and (5) repeat the numbering, verbally ... For consistency, I think I should remove those, really...

    $ LDAPS_Tmin_lapse  <dbl> 23.00694, 24.03501, 24.56563,…
    $ LDAPS_WS          <dbl> 6.818887, 5.691890, 6.138224,…
    $ LDAPS_LH          <dbl> 69.45181, 51.93745, 20.57305,…
    $ LDAPS_CC1         <dbl> 0.2339475, 0.2255082, 0.2093437,…
    $ LDAPS_CC2         <dbl> 0.2038957, 0.2517714, 0.2574694,…
    $ LDAPS_CC3         <dbl> 0.1616969, 0.1594441, 0.2040915,…
    $ LDAPS_CC4         <dbl> 0.1309282, 0.1277273, 0.1421253,…
    $ LDAPS_PPT1        <dbl> 0.0000000, 0.0000000, 0.0000000,…
    $ LDAPS_PPT2        <dbl> 0.000000, 0.000000, 0.000000,…
    $ LDAPS_PPT3        <dbl> 0.0000000, 0.0000000, 0.0000000,…
    $ LDAPS_PPT4        <dbl> 0.0000000, 0.0000000, 0.0000000,…
    $ lat               <dbl> 37.6046, 37.6046, 37.5776, 37.6450,…
    $ lon               <dbl> 126.991, 127.032, 127.058, 127.022,…
    $ DEM               <dbl> 212.3350, 44.7624, 33.3068, 45.7160,…
    $ Slope             <dbl> 2.7850, 0.5141, 0.2661, 2.5348,…
    $ `Solar radiation` <dbl> 5992.896, 5869.312, 5863.556,…
    $ Next_Tmax         <dbl> 29.1, 30.5, 31.1, 31.7, 31.2, 31.5,…
    $ Next_Tmin         <dbl> 21.2, 22.5, 23.9, 24.3, 22.5, 24.0,…

> The way we're framing the task, nearly everything in the dataset serves as a predictor. As target, we'll use `Next_Tmax`, the maximal temperature reached on the subsequent day. This means we need to remove `Next_Tmin` from the set of predictors, as it would make for too powerful of a clue. We'll do the same for `station`, the weather station id, and `Date`. This leaves us with twenty-one predictors, including measurements of actual temperature (`Present_Tmax`, `Present_Tmin`), model forecasts of various variables (`LDAPS_*`), and auxiliary information (`lat`, `lon`, and `` `Solar radiation` ``, among others).

Reviewer: "As target" feels like it's missing an article (e.g. As "a" target, as "the" target, as "our" target). If this is a common way of stating it in technical documentation, though, of course ignore my ignorance.

Author: OMG, I'm sure I'm much more ignorant than you, in that respect (you're not at all, I'm sure!).
Will change to "a"

```{r}
weather_df <- weather_df %>%
  select(-c(station, Next_Tmin, Date)) %>%
  mutate(across(.fns = scale))
```

> Note how, above, I've added a line to *standardize* the predictors. This is the "trick" I was alluding to above. To see what happens without standardization, please check out the book. (The bottom line is: You would have to call `linalg_lstsq()` with non-default arguments.)

Reviewer: Maybe note that you're using the `scale` function, just so it's clear which part you're referring to. (You could just tag that on to the end of the first sentence).

Author: will do!

  (y_true - y_pred)^2 %>%
    sum() %>%
    sqrt()
}

all_preds <- data.frame(
  b = weather_df$Next_Tmax,
  lm = fit$fitted.values
)
all_errs <- data.frame(lm = rmse(all_preds$b, all_preds$lm))
all_errs
                       
```

           lm
    1 40.8369

## Using `torch`, the quick way: `linalg_lstsq()`

Now, for a moment let's assume this was not about exploring different approaches, but getting a quick result. In `torch`, we have `linalg_lstsq()`, a function dedicated specifically to solving least-squares problems. (This is the function whose documentation I was citing, above.) Just like we did with `lm()`, we'd probably just go ahead and call it, making use of the default settings:

```{r}
x_lstsq <- linalg_lstsq(A, b)$solution

all_preds$lstsq <- as.matrix(A$matmul(x_lstsq))
all_errs$lstsq <- rmse(all_preds$b, all_preds$lstsq)

tail(all_preds)
```

                  b         lm      lstsq
    7583 -1.1380931 -1.3544620 -1.3544616
    7584 -0.8488721 -0.9040997 -0.9040993
    7585 -0.7203294 -0.9675286 -0.9675281
    7586 -0.6239224 -0.9044044 -0.9044040
    7587 -0.5275154 -0.8738639 -0.8738635
    7588 -0.7846007 -0.8725795 -0.8725792

> Predictions resemble those of `lm()` very closely -- so closely, in fact, that we may guess those tiny differences are just due to numerical errors surfacing from deep down the respective call stacks. RMSE, thus, should be equal as well:

Reviewer: 👍

```{r}
all_errs
```

           lm    lstsq
    1 40.8369 40.8369

It is; and this is a satisfying outcome. However, it only really came about due to that "trick": normalization. (Again, I have to ask you to consult the book for details.)

Now, let's explore what we can do without using `linalg_lstsq()`.

## Least squares (I): The normal equations

> We start by stating the goal. Given a matrix, $\mathbf{A}$, that holds features in its columns and observations in its rows, and a vector of observed outcomes, $\mathbf{b}$, we want to find regression coefficients, one for each feature, that allow to approximate $\mathbf{b}$ as well as possible. Call the vector of regression coefficients $\mathbf{x}$. To obtain it, we need to solve a simultaneous system of equations, that in matrix notation appears as

Reviewer: ```suggestion
We start by stating the goal. Given a matrix, $\mathbf{A}$, that holds features in its columns and observations in its rows, and a vector of observed outcomes, $\mathbf{b}$, we want to find regression coefficients, one for each feature, that allow us to approximate $\mathbf{b}$ as well as possible. Call the vector of regression coefficients $\mathbf{x}$. To obtain it, we need to solve a simultaneous system of equations, that in matrix notation appears as
```

$$
\mathbf{Ax} = \mathbf{b}
$$

If $\mathbf{b}$ were a square, invertible matrix, the solution could directly be computed as $\mathbf{x} = \mathbf{A}^{-1}\mathbf{b}$. This will hardly ever be possible, though; we'll (hopefully) always have more observations than predictors. Another approach is needed. It directly starts from the problem statement.

> When we use the columns of $\mathbf{A}$ to approximate $\mathbf{b}$, that approximation necessarily is in the column space of $\mathbf{A}$. $\mathbf{b}$, on the other hand, normally won't be. We want those two to be as close as possible; in other words, we want to minimize the distance between them. Choosing the 2-norm for the distance, this yields the objective

Reviewer: ```suggestion
When we use the columns of $\mathbf{A}$ to approximate $\mathbf{b}$, that approximation necessarily is in the column space of $\mathbf{A}$. $\mathbf{b}$, on the other hand, normally won't be. We want those two to be as close as possible. In other words, we want to minimize the distance between them. Choosing the 2-norm for the distance, this yields the objective
```

## Least squares (II): Cholesky decomposition

In Cholesky decomposition, a matrix is factored into two triangular matrices of the same size, with one being the transpose of the other. This commonly is written either

$$
\mathbf{A} = \mathbf{L} \mathbf{L}^T
$$ or

$$
\mathbf{A} = \mathbf{R}^T\mathbf{R}
$$

Here symbols $\mathbf{L}$ and $\mathbf{R}$ denote lower-triangular and upper-triangular matrices, respectively.

> For Cholesky decomposition to be possible, a matrix has to be both symmetric and positive definite. These are pretty strong conditions, ones that will not often be fulfilled in practice. In our case, $\mathbf{A}$ is not symmetric; this immediately implies we have to operate on $\mathbf{A}^T\mathbf{A}$ instead. And since $\mathbf{A}$ already is positive definite, we know that $\mathbf{A}^T\mathbf{A}$ is, as well.

Reviewer: ```suggestion
For Cholesky decomposition to be possible, a matrix has to be both symmetric and positive definite. These are pretty strong conditions, ones that will not often be fulfilled in practice. In our case, $\mathbf{A}$ is not symmetric. This immediately implies we have to operate on $\mathbf{A}^T\mathbf{A}$ instead. And since $\mathbf{A}$ already is positive definite, we know that $\mathbf{A}^T\mathbf{A}$ is, as well.
```

With $\mathbf{L}\mathbf{L}^T$, the point is that $\mathbf{L}$ is triangular, and when that's the case the linear system can be solved by simple substitution. That is best visible with a tiny example:

$$
\begin{bmatrix}
  1 & 0 & 0\\
  2 & 3 & 0\\
  3 & 4 & 1
\end{bmatrix}
\begin{bmatrix}
  x1\\
  x2\\
  x3
\end{bmatrix}
=
\begin{bmatrix}
  1\\
  11\\
  15
\end{bmatrix}
$$

Starting in the top row, we immediately see that $x1$ equals $1$; and once we know *that* it is straightforward to calculate, from row two, that $x2$ must be $3$. The last row then tells us that $x3$ must be $0$.

In code, `torch_triangular_solve()` is used to efficiently compute the solution to a linear system of equations where the matrix of predictors is lower- or upper-triangular. An additional requirement is for the matrix to be symmetric -- but that condition we already had to satisfy in order to be able to use Cholesky factorization.

> By default, `torch_triangular_solve()` expects the matrix to be upper- (not lower-)triangular; but there is a function parameter, `upper`, that lets us correct that expectation. The return value is a list, and its first item contains the desired solution. To illustrate, here is `torch_triangular_solve()`, applied to the toy example we manually solved above:

Reviewer: ```suggestion
By default, `torch_triangular_solve()` expects the matrix to be upper- (not lower-) triangular; but there is a function parameter, `upper`, that lets us correct that expectation. The return value is a list, and its first item contains the desired solution. To illustrate, here is `torch_triangular_solve()`, applied to the toy example we manually solved above:
```

$$
\begin{aligned}
\mathbf{A}\mathbf{x} &= \mathbf{b}\\
\mathbf{Q}\mathbf{R}\mathbf{x} &= \mathbf{b}\\
\mathbf{R}\mathbf{x} &= \mathbf{Q}^T\mathbf{b}\\
\end{aligned}
$$

In `torch`, `linalg_qr()` gives us the matrices $\mathbf{Q}$ and $\mathbf{R}$.

```{r}
c(Q, R) %<-% linalg_qr(A)
```

On the right side, we used to have a "convenience variable" holding $\mathbf{A}^T\mathbf{b}$ ; here, we skip that step, and instead, do something "immediately useful": move $\mathbf{Q}$ to the other side.

```{r}
Qtb <- Q$t()$matmul(b)
```

The only remaining step now is to solve the remaining triangular system.

```{r}
x <- torch_triangular_solve(Qtb$unsqueeze(2), R)[[1]]

all_preds$qr <- as.matrix(A$matmul(x))
all_errs$qr <- rmse(all_preds$b, all_preds$qr)
all_errs[1, -c(5,7)]
```

           lm   lstsq     neq    chol      lu      qr
    1 40.8369 40.8369 40.8369 40.8369 40.8369 40.8369 

> By now, you'll be expecting for me to end this section saying "there is also a dedicated solver in `torch`/`torch_linalg`, namely ..."). Well, not literally, no; but effectively, yes. If you call `linalg_lstsq()` passing `driver = "gels"`, it is QR factorization that will be used.

Reviewer: ```suggestion
By now, you'll be expecting for me to end this section saying "there is also a dedicated solver in `torch`/`torch_linalg`, namely ..."). Well, not literally, no; but effectively, yes. If you call `linalg_lstsq()` passing `driver = "gels"`, QR factorization will be used.
```
